File: music-infrastructure/src/main/java/com/music/sale/persistence/common/JsonConverter.java
// Copyright (C) 2024 Your Name or Company
package com.music.sale.persistence.common;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.persistence.AttributeConverter;
import jakarta.persistence.Converter;
import java.util.HashMap;
import java.util.Map;

@Converter
public class JsonConverter implements AttributeConverter<Map<String, Object>, String> {
  private final ObjectMapper objectMapper = new ObjectMapper();

  @Override
  public String convertToDatabaseColumn(Map<String, Object> attribute) {
    try {
      return attribute != null ? objectMapper.writeValueAsString(attribute) : null;
    } catch (Exception e) {
      System.out.println("Error serializing to JSON: " + e.getMessage());
      return null;
    }
  }

  @Override
  public Map<String, Object> convertToEntityAttribute(String dbData) {
    try {
      if (dbData == null || dbData.isBlank()) {
        return new HashMap<>();
      }

      return objectMapper.readValue(dbData, new TypeReference<HashMap<String, Object>>() {});
    } catch (Exception e) {
      System.out.println("Error deserializing JSON: " + e.getMessage() + ", data: " + dbData);
      return new HashMap<>();
    }
  }
}

File: src/main/kotlin/com/music/sale/adapter/persistence/common/JsonConverter.kt
// Copyright (C) 2024 Your Name or Company
package com.music.sale.adapter.persistence.common

import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.persistence.AttributeConverter
import jakarta.persistence.Converter

@Converter
class JsonConverter : AttributeConverter<Map<String, Any>, String> {
    private val objectMapper = ObjectMapper()

    override fun convertToDatabaseColumn(attribute: Map<String, Any>?): String? {
        return try {
            attribute?.let { objectMapper.writeValueAsString(it) }
        } catch (e: Exception) {
            println("Error serializing to JSON: ${e.message}")
            null
        }
    }

    override fun convertToEntityAttribute(dbData: String?): Map<String, Any>? {
        return try {
            if (dbData.isNullOrBlank()) {
                return emptyMap()
            }
            
            objectMapper.readValue(dbData, object : TypeReference<HashMap<String, Any>>() {})
        } catch (e: Exception) {
            println("Error deserializing JSON: ${e.message}, data: $dbData")
            // JSON 파싱 실패 시 빈 맵 반환
            emptyMap()
        }
    }
}
